codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customizations/InterceptorConfigCustomization.kt

```diff
@@ -0,0 +1,185 @@
+        "SharedInterceptor" to interceptors.resolve("SharedInterceptor"),
+    )
+
+    override fun section(section: ServiceConfig) =
+        writable {
+            when (section) {
+                is ServiceConfig.ConfigStruct -> rustTemplate(
+                    """
+                    // TODO(enableNewSmithyRuntime): Unused until we completely switch to the orchestrator
+                    ##[allow(dead_code)]
+                    pub(crate) interceptors: Vec<#{SharedInterceptor}>,
+                    """,
+                    *codegenScope,
+                )
+
+                is ServiceConfig.BuilderStruct ->
+                    rustTemplate(
+                        """
+                        interceptors: Vec<#{SharedInterceptor}>,
+                        """,
+                        *codegenScope,
+                    )
+
+                is ServiceConfig.ConfigImpl -> writable {
+                    rustTemplate(
+                        """
+                        /// Returns interceptors currently registered by the user
+                        pub fn interceptors(&self) -> impl Iterator<Item = &#{SharedInterceptor}> + '_ {
+                            self.interceptors.iter()
+                        }
+                        """,
+                        *codegenScope,
+                    )
+                }
+
+                ServiceConfig.BuilderImpl ->
+                    rustTemplate(
+                        """
+                        ##[doc(hidden)]
+                        /// Sets an [`Interceptor`](#{Interceptor}) that runs at specific stages of the request execution pipeline.
+                        ///
+                        /// Interceptors targeted at a certain stage are executed according to the pre-defined priority.
+                        /// SDK provides the default set of interceptors. An interceptor configured by this method
+                        /// will run after those default interceptors.
+                        ///
+                        /// ## Examples
+                        /// ```no_run
+                        /// ## ##[cfg(test)]
+                        /// ## mod tests {
+                        /// ## ##[test]
+                        /// ## fn example() {
+                        /// use aws_http::user_agent::AwsUserAgent;
+                        /// use aws_smithy_runtime_api::client::interceptors::{Interceptor, InterceptorContext};
+                        /// use aws_smithy_runtime_api::config_bag::ConfigBag;
+                        /// use http::header::USER_AGENT;
+                        /// use http::{HeaderName, HeaderValue};
+                        /// use $moduleUseName::config::Config;
+                        ///
+                        /// ##[derive(Debug)]
+                        /// struct TestUserAgentInterceptor;
+                        /// impl Interceptor for TestUserAgentInterceptor {
```

Comment:
This example is for generic clients, so it can't reference any AWS code like `AwsUserAgent`, and shouldn't set a `x-amz-user-agent` header.

Reply:
That did cross my mind. I'll come up with a different example.

Reply:
Updated the example in 5583680.